File: src/com/train/algorithm/list/ReverseList.java
package com.train.algorithm.list;


import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

public class ReverseList {
    public class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ListNode reverseList(ListNode head){
        if (null == head || null == head.next){
            return head;
        }
        ListNode pre = null;
        while (null!=head){
            ListNode temp = head;
            head = head.next;
            temp.next = pre;
            pre = temp;
        }
        return pre;
    }

    public ListNode reverseList1(ListNode head){
        if (null == head || null == head.next){
            return head;
        }
        List<Integer> list = new LinkedList<>();
        ListNode tempNode = head;
        while (null!=tempNode){
            list.add(tempNode.val);
            tempNode = tempNode.next;
        }
        Collections.reverse(list);
        tempNode = head;
        while (null != tempNode){
            tempNode.val = list.remove(0);
            tempNode = tempNode.next;
        }
        return head;
    }

    // leetcode-25
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode h = new ListNode(0);
        h.next = head;
        ListNode pre = h;
        while(head!=null){
            ListNode tail = pre;
            for (int i = 0;i < k;i++ ) {
                tail = tail.next;
                if (tail == null) {
                    return h.next;
                }
            }
            ListNode nex = tail.next;
            ListNode[] revers = reverse(head,tail);
            head = revers[0];
            tail = revers[1];
            pre.next = head;
            tail.next = nex;
            pre = tail;
            head = tail.next;
        }
        return h.next;
    }
    public ListNode[] reverse(ListNode head,ListNode tail){
        ListNode pre = tail.next;
        ListNode p = head;
        while(pre!=tail){
            ListNode temp = p.next;
            p.next = pre;
            pre = p;
            p = temp;
        }
        return new ListNode[]{tail,head};
    }
}

File: src/com/train/algorithm/list/implementInJS/ReverseList.js
// 反转链表
// 剑指offer 24
var reverseList = function(head) {
	if(head == null || head.next == null){
		return head;
	}
	let pre = null;
	let cur = head;
	while(cur != null){
		const next = cur.next;
		cur.next = pre;
		pre = cur;
		cur = next;
	}
	return pre;
};

// 剑指offer 24
var reverseList = function(head) {
	if(head == null || head.next == null){
		return head;
	}
	const newHead = reverseList(head.next);
	head.next.next = head;
	head.next = null;
	return newHead;
};

// leetcode-25
const reverse = function(head,tail){
	let pre = tail.next;
	let p = head;
	while(pre != tail){
		const temp = p.next;
		p.next = pre;
		pre = p;
		p = temp;
	}
	return [tail,head];
}
var reverseKGroup = function(head, k) {
	const h = new ListNode(0);
	h.next = head;
	let pre = h;
	while(head!=null){
		let tail = pre;
		for(let i = 0;i < k;i++){
			tail = tail.next;
			if(tail == null){
				return h.next;
			}
		}
		let nex = tail.next;
		[head,tail] = reverse(head,tail);
		pre.next = head;
		tail.next = nex;
		pre = tail;
		head = tail.next;
	}
	return h.next;
};